File: backend-python/modules/payments/public_api/service.py
"""ONLY interface other modules/apps may import from the payments module."""
from modules.payments.internal.payment_service import PaymentService

class PaymentsPublicApi:
    def __init__(self):
        self._svc = PaymentService()

    def authorize(self, order_id: str, amount_cents: int, token: str) -> dict:
        return self._svc.authorize(order_id, amount_cents, token)

    def capture(self, payment_id: str) -> dict | None:
        return self._svc.capture(payment_id)

    def refund(self, payment_id: str, amount_cents: int) -> dict | None:
        return self._svc.refund(payment_id, amount_cents)

File: backend-golang/modules/payments/internal/service.go
// Package internal holds payments business logic hidden from other modules.
package internal

import "sync"

type Payment struct {
	PaymentID     string `json:"payment_id"`
	OrderID       string `json:"order_id"`
	Status        string `json:"status"`
	AmountCents   int64  `json:"amount_cents"`
	RefundedCents int64  `json:"refunded_cents,omitempty"`
}

var (
	mu       sync.Mutex
	payments = map[string]*Payment{}
)

func Authorize(paymentID, orderID string, amountCents int64) Payment {
	mu.Lock()
	defer mu.Unlock()
	p := &Payment{PaymentID: paymentID, OrderID: orderID, Status: "AUTHORIZED", AmountCents: amountCents}
	payments[paymentID] = p
	return *p
}

func Capture(paymentID string) (Payment, bool) {
	mu.Lock()
	defer mu.Unlock()
	p, ok := payments[paymentID]
	if !ok || p.Status != "AUTHORIZED" {
		return Payment{}, false
	}
	p.Status = "CAPTURED"
	return *p, true
}

func Refund(paymentID string, amountCents int64) (Payment, bool) {
	mu.Lock()
	defer mu.Unlock()
	p, ok := payments[paymentID]
	if !ok || p.Status != "CAPTURED" {
		return Payment{}, false
	}
	p.Status = "REFUNDED"
	p.RefundedCents = amountCents
	return *p, true
}
